Mathlib/Topology/ContinuousFunction/CompactlySupported.lean

```diff
@@ -0,0 +1,551 @@
+  rfl
+
+theorem copy_eq (f : C_c(α, β)) (f' : α → β) (h : f' = f) : f.copy f' h = f :=
+  DFunLike.ext' h
+
+theorem eq_of_empty [IsEmpty α] (f g : C_c(α, β)) : f = g :=
+  ext <| IsEmpty.elim ‹_›
+
+/-- A continuous function on a compact space has automatically compact support. -/
+@[simps]
+def ContinuousMap.liftCompactlySupported [CompactSpace α] : C(α, β) ≃ C_c(α, β) where
+  toFun f :=
+    { toFun := f
+      continuous_toFun := f.continuous
+      hasCompactSupport' := hasCompactSupport_of_compactSpace f }
+  invFun f := f
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+/-- A continuous function on a compact space has automatically compact support. This is not an
+instance to avoid type class loops. -/
+lemma _root_.compactlySupportedContinuousMapClass.ofCompact (G : Type*) [FunLike G α β]
+    [ContinuousMapClass G α β] [CompactSpace α] : CompactlySupportedContinuousMapClass G α β where
+  map_continuous := map_continuous
+  hasCompactSupport := by
+    intro f
+    exact hasCompactSupport_of_compactSpace f
+
+end Basics
+
+/-! ### Algebraic structure
+
+Whenever `β` has the structure of continuous addtive monoid and a compatible topological structure,
+then `C_c(α, β)` inherits a corresponding algebraic structure. The primary exception to this is that
+`C_c(α, β)` will not have a multiplicative identity.
+-/
+
+section AlgebraicStructure
+
+variable [TopologicalSpace β] (x : α)
+
+instance [Zero β] : Zero C_c(α, β) where
+  zero := { toFun := (0 : C(α, β))
+            continuous_toFun := (0 : C(α, β)).2
+            hasCompactSupport' := by
+              rw [HasCompactSupport, tsupport]
+              simp only [ContinuousMap.coe_zero, Function.support_zero', closure_empty,
+                isCompact_empty] }
+
+instance [Zero β] : Inhabited C_c(α, β) :=
+  ⟨0⟩
+
+@[simp]
+theorem coe_zero [Zero β] : ⇑(0 : C_c(α, β)) = 0 :=
+  rfl
+
+theorem zero_apply [Zero β] : (0 : C_c(α, β)) x = 0 :=
+  rfl
+
+instance [MulZeroClass β] [ContinuousMul β] : Mul C_c(α, β) :=
+  ⟨fun f g => ⟨f * g, HasCompactSupport.mul_left g.2⟩⟩
+
+@[simp]
+theorem coe_mul [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : ⇑(f * g) = f * g :=
+  rfl
+
+theorem mul_apply [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : (f * g) x = f x * g x :=
+  rfl
+
+/-- the product of `f : F` assuming `ContinuousMapClass F α γ` and `ContinuousSMul γ β` and
+`g : C_c(α, β)` is in `C_c(α, β)` -/
+instance [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
+    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] : SMul F C_c(α, β) where
+  smul f g :=
+    ⟨⟨fun x ↦ f x • g x, (map_continuous f).smul g.continuous⟩, g.hasCompactSupport'.smul_left⟩
+
+@[simp]
+theorem coe_smulc [MulZeroClass β] [ContinuousMul β] (f : C(α, β)) (g : C_c(α, β)) :
+    ⇑(f • g) = f * g :=
+  rfl
+
+@[simp]
+theorem smulc_apply [MulZeroClass β] [ContinuousMul β] (f : C(α, β)) (g : C_c(α, β)) :
+    (f • g) x = f x * g x :=
+  rfl
```

Comment:
These should rather be stated in the general setting of the instance above.

Reply:
What do you mean by "in the general setting"? They are stated just below the definition of the instance `SMul F C_c(α, β)`.

Reply:
Here, these lemmas are stated for `(f : C(α, β)) (g : C_c(α, β))`. They should be for `(f : F) (g : C_c(α, β))`, where `F` is like in the instance (and the multiplication will be replaced by •, probably).

Reply:
I see. `coe_smulc` now becomes `⇑(f • g) = fun x => f x • g x`, as this seems to be the only thing that makes sense. Is this useful?